notebooks/vision/humans_or_horses.livemd

```diff
@@ -1,258 +0,0 @@
-    |> Axon.conv(64, kernel_size: {3, 3}, activation: :relu)
-    |> Axon.max_pool(kernel_size: {2, 2})
-    |> Axon.flatten()
-    |> Axon.dropout(rate: 0.5)
-    |> Axon.dense(512, activation: :relu)
-    |> Axon.dense(2, activation: :softmax)
-  end
-end
-```
-
-## Training
-
-Creating a training loop starts with setting a loss function to categorical cross-entropy and optimizer of your choice. We also set the log option to 1 to keep track of loss and metrics during training. Next, we choose accuracy as a metric. Finally, we pass the number of epochs and set the number of iterations per epoch to utilize almost all of the images.
-
-```elixir
-defmodule HorsesHumans.Training do
-  def train_model(model, data, optimizer, epochs) do
-    model
-    |> Axon.Loop.trainer(:categorical_cross_entropy, optimizer, log: 1)
-    |> Axon.Loop.metric(:accuracy)
-    |> Axon.Loop.run(data, %{}, epochs: epochs, iterations: Config.batches_per_epoch())
-  end
-end
-```
-
-## Model evaluation
-
-Now we can build and train our model. We start with passing input shape and indices to data transposition. Next, we set the optimizer to adam with a learning rate of 1.0e-4. In the next step, we define an optimizer but with gradient centralization. It is a technique with a similar purpose to batch normalization. For each loss gradient, we subtract a mean value to have a gradient with mean equal to zero. This process prevents us from exploding gradients.
-
-```elixir
-defmodule HorsesHumans.ModelTraining do
-  def run() do
-    model = HorsesHumans.Model.build_model({nil, 300, 300, 4}, [2, 0, 1]) |> IO.inspect()
-    optimizer = Axon.Optimizers.adam(1.0e-4)
-    centralized_optimizer = Axon.Updates.compose(Axon.Updates.centralize(), optimizer)
-
-    data = HorsesHumans.DataProcessing.data_stream()
-
-    IO.write("\n\nTraining model without gradient centralization\n\n")
-    HorsesHumans.Training.train_model(model, data, optimizer, 2)
-
-    IO.write("\n\nTraining model with gradient centralization\n\n")
-    HorsesHumans.Training.train_model(model, data, centralized_optimizer, 2)
-  end
-end
-```
-
-```elixir
-HorsesHumans.ModelTraining.run()
-```
-
-## Conclusions
-
-We have an accuracy of almost 1.0.  That is great right?  Hmmm. Not so great.  With this notebook we've demonstrated the ability for a neural net model to memorize a dataset.  The model knows how to distinguish each image with the kind of image it is.  Training a model is tricky.  It may learn unintended characteristics.  Do all of the horses have something in the background that the model can make it's decision on?
-
-For useful models, we need to compare the model against images that it hasn't seen.  Comparison against unseen input data measures the generalization of the model with respect to the unseen images.  Using the model against further unseen images will help with measure the bias of the model.  How well will this model perform against a real horse or pony?
```

Comment:
I removed this, I believe this was actually a bug. The target class was assigned based on `String.contains?(filename, "horses")`, checking against the full path, and I think the path would always include `"horses"` as in `"data/humans_or_horses/{horses,humans}/*"`. So we would train the model on a single class :)